lib/srv/server/ec2_watcher.go

```diff
@@ -146,26 +188,98 @@ func newEC2InstanceFetcher(cfg ec2FetcherConfig) *ec2InstanceFetcher {
+	insts := EC2Instances{
+		Region:       f.Region,
+		DocumentName: f.DocumentName,
+		Parameters:   f.Parameters,
+		Rotation:     rotation,
+	}
+	for _, node := range nodes {
+		if node.GetSubKind() != types.SubKindOpenSSHNode {
+			continue
+		}
+		region, ok := node.GetLabel(types.AWSInstanceRegion)
+		if !ok || region != f.Region {
+			continue
+		}
+		instID, ok := node.GetLabel(types.AWSInstanceIDLabel)
+		if !ok {
+			continue
+		}
+		accountID, ok := node.GetLabel(types.AWSAccountIDLabel)
+		if !ok {
+			continue
+		}
+
+		_, ok = f.foundInstances[foundInstancesKey{accountID: accountID, instanceID: instID}]
+		if !ok {
+			continue
+		}
+		if insts.AccountID != "" {
+			insts.AccountID = accountID
+		}
+
+		insts.Instances = append(insts.Instances, EC2Instance{
+			InstanceID: instID,
+		})
+	}
+
+	if len(insts.Instances) == 0 {
+		return nil, trace.NotFound("no ec2 instances found")
+	}
+
+	return chunkInstances(insts), nil
+}
+
+func chunkInstances(insts EC2Instances) []Instances {
+	var instColl []Instances
+	for i := 0; i < len(insts.Instances); i += awsEC2APIChunkSize {
+		end := i + awsEC2APIChunkSize
+		if end > len(insts.Instances) {
+			end = len(insts.Instances)
+		}
+		inst := EC2Instances{
+			AccountID:    insts.AccountID,
+			Region:       insts.Region,
+			DocumentName: insts.DocumentName,
+			Parameters:   insts.Parameters,
+			Instances:    insts.Instances[i:end],
+			Rotation:     insts.Rotation,
+		}
+		instColl = append(instColl, Instances{EC2Instances: &inst})
+	}
+	return instColl
+}
+
 // GetInstances fetches all EC2 instances matching configured filters.
-func (f *ec2InstanceFetcher) GetInstances(ctx context.Context) ([]Instances, error) {
+func (f *ec2InstanceFetcher) GetInstances(ctx context.Context, rotation bool) ([]Instances, error) {
 	var instances []Instances
+	f.foundInstances = make(map[foundInstancesKey]struct{}) // clear found instances as they're refreshed bellow
 	err := f.EC2.DescribeInstancesPagesWithContext(ctx, &ec2.DescribeInstancesInput{
 		Filters: f.Filters,
 	},
 		func(dio *ec2.DescribeInstancesOutput, b bool) bool {
-			const chunkSize = 50 // max number of instances SSM will send commands to at a time
 			for _, res := range dio.Reservations {
-				for i := 0; i < len(res.Instances); i += chunkSize {
-					end := i + chunkSize
+				for i := 0; i < len(res.Instances); i += awsEC2APIChunkSize {
+					end := i + awsEC2APIChunkSize
 					if end > len(res.Instances) {
 						end = len(res.Instances)
 					}
+					ownerID := aws.StringValue(res.OwnerId)
 					inst := EC2Instances{
-						AccountID:    aws.StringValue(res.OwnerId),
+						AccountID:    ownerID,
 						Region:       f.Region,
 						DocumentName: f.DocumentName,
-						Instances:    res.Instances[i:end],
+						Instances:    ToEC2Instances(res.Instances[i:end]),
 						Parameters:   f.Parameters,
+						Rotation:     rotation,
+					}
+					for _, ec2inst := range res.Instances[i:end] {
+						f.foundInstances[foundInstancesKey{
```

Comment:
Is `f.foundInstances` a shared state map? If so, it should be protected from concurrent access. I would make helper methods for adding items to this map and checking for existence.

Reply:
I would, yeah.